### PR TITLE
[Review Only] Fuzzer repro. Will separate to multiple PRs for landing. This is only for reviewing

### PR DESCRIPTION
**Pull request by tanjialiang**
Extract generateFilePath() out to be a single place to remove duplicate code.

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang First commit looks good me. Would you like to extract it into a PR?

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang 2nd commit looks great.

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 1]
```diff
@@ -73,6 +74,13 @@ DEFINE_bool(
     false,
     "Enable testing of function signatures with variadic arguments.");
 
+DEFINE_string(
+    repro_persist_path,
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp]: I wonder if there is a better name for this flag.

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang A few initial comments on the 3rd commit.

File: velox/expression/tests/ExpressionVerifier.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ private:
+  void printRowVector(const RowVectorPtr& rowVector);
+  void compareExceptions(const VeloxException& ve1, const VeloxException& ve2);
+
+  // Called if at least one of the ptrs has an exception.
+  void compareExceptions(
+      std::exception_ptr commonPtr,
+      std::exception_ptr simplifiedPtr);
+
+  void compareVectors(const VectorPtr& vec1, const VectorPtr& vec2);
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.h:54]: left and right, maybe?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/tests/ExpressionVerifier.h"
+#include "velox/expression/Expr.h"
+#include "velox/vector/VectorSaver.h"
+
+DEFINE_bool(
+    disable_constant_folding,
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.cpp:22]: Why do we need this?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    false,
+    "Disable constant-folding in the common evaluation path.");
+
+DEFINE_string(
+    repro_persist_path,
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.cpp:27]: Why do we need this here? Can we pass these as arguments?

File: velox/vector/VectorSaver.h [hunk 5]
```diff
@@ -34,12 +37,25 @@ TypePtr restoreType(std::istream& in);
 /// output stream. The serialiation preserved encoding.
 void saveVector(const BaseVector& vector, std::ostream& out);
 
+/// Serializes the vector into binary format and writes it to a new file in
+/// 'filePath'. The serialiation preserved encoding. Exceptions will be thrown
```
Comment (mbasmanova) [inline velox/vector/VectorSaver.h:41]: typo: "serialiation preserved"

File: velox/vector/VectorSaver.h [hunk 6]
```diff
@@ -34,12 +37,25 @@ TypePtr restoreType(std::istream& in);
...
+/// if any error occurs while writing
```
Comment (mbasmanova) [inline velox/vector/VectorSaver.h:42]: add period at the end of the sentence ; ditto other places

File: velox/vector/VectorSaver.h [hunk 7]
```diff
@@ -22,6 +22,9 @@ namespace facebook::velox {
 // Serialization format used by the following function is documented at
 // https://facebookincubator.github.io/velox/develop/debugging/vector-saver.html
 
+static constexpr folly::StringPiece kVectorPathPrefix = "vector";
```
Comment (mbasmanova) [inline velox/vector/VectorSaver.h:25]: I'm not sure it makes sense to have these constants. The callers should decide for themselves how to name the files. In particular, I don't think there is a use case for kStringPathPrefix.

File: velox/expression/Expr.cpp [hunk 8]
```diff
@@ -334,38 +334,34 @@ class ExprExceptionContext {
...
     try {
-      auto sql = expr_->toSql();
-      std::ofstream outputFile(sqlPath.value(), std::ofstream::binary);
-      outputFile.write(sql.data(), sql.size());
-      outputFile.close();
-      sqlPath_ = sqlPath.value();
-    } catch (...) {
-      sqlPath_ =
-          fmt::format("Failed to save expression SQL to {}.", sqlPath.value());
+      auto sqlPathOpt = generateFilePath(basePath, kStringPathPrefix.data());
```
Comment (mbasmanova) [inline velox/expression/Expr.cpp:354]: kStringPathPrefix -> "sql"

File: velox/expression/Expr.cpp [hunk 9]
```diff
@@ -334,38 +334,34 @@ class ExprExceptionContext {
...
+      if (!sqlPathOpt.has_value()) {
+        // dataPath_ is used to deliver error message.
```
Comment (mbasmanova) [inline velox/expression/Expr.cpp:356]: Why is this change? Let's use sqlPath_

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 10]
```diff
@@ -73,6 +74,12 @@ DEFINE_bool(
...
+    "",
+    "Base directory path for persistence of data and SQL when fuzzer fails "
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp:80]: "Base directory path" -> "Directory path"

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 11]
```diff
@@ -73,6 +74,12 @@ DEFINE_bool(
...
+    "for future reproduction. Empty string will disable this feature.");
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp:81]: "will disable" -> "disables" or "Use empty string to disable this feature."

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 12]
```diff
@@ -580,22 +649,33 @@ class ExpressionFuzzer {
       }
     }
 
+    // Store data and expression in case of reproduction.
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp]: "in case of reproduction" -> "for repro..."

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 13]
```diff
@@ -687,18 +774,34 @@ class ExpressionFuzzer {
...
 
         plan = std::make_shared<core::CallTypedExpr>(
             plan->type(), std::vector<core::TypedExprPtr>{plan}, "try");
 
         // At this point, the function throws if anything goes wrong.
-        executeExpression(plan, rowVector, false);
+        executeExpression(
+            plan,
+            rowVector,
+            resultVector ? BaseVector::copy(*resultVector) : nullptr,
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp:532]: By this time resultVector may have changed. Looks like before we would create new results vector on each invocation, which doesn't seem right. I wonder if proper way is to make a copy of the result vector before calling executeExpression for the first time, then use that copy here.

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang Thank you for iterating on this diff. Here are some further comments on 2nd commit.

File: velox/expression/tests/ExpressionVerifier.h [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      const RowVectorPtr& rowVector,
+      VectorPtr&& resultVector,
+      bool canThrow);
+
+ private:
+  void persistReproInfo(
+      const VectorPtr& inputVector,
+      const VectorPtr& resultVector,
+      const std::string& sql);
+  void printRowVector(const RowVectorPtr& rowVector);
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.h:57]: add empty lines between methods

File: velox/expression/tests/ExpressionVerifier.h [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+  // Called if at least one of the ptrs has an exception.
+  void compareExceptions(
+      std::exception_ptr commonPtr,
+      std::exception_ptr simplifiedPtr);
+
+  void compareVectors(const VectorPtr& left, const VectorPtr& right);
+
+ private:
+  core::ExecCtx* FOLLY_NONNULL execCtx_;
+  ExpressionVerifierOptions options_;
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.h:71]: Can this be const?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+  ss << "Persisted input at '" << inputPath;
+  if (resultVector) {
+    ss << "' and result at '" << resultPath;
+  }
+  ss << "' and sql at '" << sqlPath << "'";
+  LOG(INFO) << ss.str();
+}
+
+
+void ExpressionVerifier::printRowVector(const RowVectorPtr& rowVector) {
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.cpp:191]: this method could be a free function in the anonymous namespace in the .cpp file

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+  LOG(INFO) << "RowVector contents (" << rowVector->type()->toString() << "):";
+
+  for (vector_size_t i = 0; i < rowVector->size(); ++i) {
+    LOG(INFO) << "\tAt " << i << ": " << rowVector->toString(i);
+  }
+}
+
+void ExpressionVerifier::compareExceptions(
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.cpp:199]: same

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+       left.errorCode() == "UNSUPPORTED"));
+  VELOX_CHECK_EQ(left.errorSource(), right.errorSource());
+  VELOX_CHECK_EQ(left.exceptionName(), right.exceptionName());
+  if (left.message() != right.message()) {
+    LOG(WARNING) << "Two different VeloxExceptions were thrown:\n\t"
+                 << left.message() << "\nand\n\t" << right.message();
+  }
+}
+
+void ExpressionVerifier::compareExceptions(
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.cpp:224]: same

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+        LOG(WARNING) << "Two different std::exceptions were thrown:\n\t"
+                     << e1.what() << "\nand\n\t" << e2.what();
+      }
+      return;
+    }
+  }
+  VELOX_FAIL("Got two incompatible exceptions.");
+}
+
+void ExpressionVerifier::compareVectors(
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.cpp:264]: same

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 20]
```diff
@@ -65,14 +65,14 @@ DEFINE_bool(
     "Retry failed expressions by wrapping it using a try() statement.");
 
 DEFINE_bool(
-    disable_constant_folding,
+    enable_variadic_signatures,
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzer.cpp:68]: unrelated change?

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang Some comments on the 3rd commit.

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 21]
```diff
@@ -16,6 +16,7 @@
 
 #include "velox/expression/tests/ExpressionVerifier.h"
 #include "velox/expression/Expr.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionVerifier.cpp:19]: this change might belong to a previous commit

File: velox/expression/tests/ExpressionFuzzerTest.cpp [hunk 22]
```diff
@@ -51,8 +51,7 @@ int main(int argc, char** argv) {
       // (since HLL is a user defined type), and end up trying to use
       // cardinality passing a VARBINARY (since HLL's implementation uses an
       // alias to VARBINARY).
-      "cardinality",
-      "neq"};
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionFuzzerTest.cpp:55]: unintended change?

File: velox/expression/tests/ExpressionRunner.cpp [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+    "Path for result vector to restore from disk. This is optional for "
+    "on-disk reproduction. Don't set if the initial repro result vector is "
+    "nullptr");
+
+// TODO(jtan6): Support common mode and simplified mode.
+DEFINE_string(
+    mode,
+    "verify",
+    "Mode for expression runner: \n"
+    "verify: run expression and verify between common and simplified path.\n"
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionRunner.cpp:55]: - path -> paths
- verify -> compare results

File: velox/expression/tests/ExpressionRunner.cpp [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    "common: run expression only in common path.\n"
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionRunner.cpp:56]: Should we make this the default?

Comment (mbasmanova) [inline velox/expression/tests/ExpressionRunner.cpp:56]: in -> using

File: velox/expression/tests/ExpressionRunner.cpp [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+    "simplified: run expression only in simplified path.");
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionRunner.cpp:57]: in -> using

File: velox/expression/tests/ExpressionRunner.cpp [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+using namespace facebook::velox;
+
+int main(int argc, char** argv) {
```
Comment (mbasmanova) [inline velox/expression/tests/ExpressionRunner.cpp:61]: Should we extract the logic into a helper class and add a unit test. Otherwise, this code can get broken without us noticing.

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang Some further comments on the last commit.

File: velox/expression/Expr.cpp (hunk 9 above)
Comment (tanjialiang) [reply velox/expression/Expr.cpp:356]: I somehow remember the original implementation uses dataPath for error message. But since changes have been made and it does not make sense to have datapath to store the error message anymore. This change is wrong here in the commit. I'll revert it back

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 13 above)
Comment (tanjialiang) [reply velox/expression/tests/ExpressionFuzzer.cpp:532]: I think the current implementation does exactly what you described. 
resultVector is never passed in executeExpression(). Only its copies were passed in so it should stay unchanged.
